lib/matplotlib/cbook.py

```diff
@@ -783,21 +783,27 @@ class Grouper:
     """
 
     def __init__(self, init=()):
-        self._mapping = weakref.WeakKeyDictionary(
-            {x: weakref.WeakSet([x]) for x in init})
+        self._count = itertools.count()
+        # For each item, we store (order_in_which_item_was_seen, group_of_item), which
+        # lets __iter__ and get_siblings return items in the order in which they have
+        # been seen.
```

Comment:
Is this necessary? If I understand this correctly, you just encode the insertion order into the WeakKeyDictionary. Can't we rely on order preservation there? I've not tested this thoroughly but the internal data format `WeakKeyDictionary.data` is a plain dict, which preserves order. So it should work as is. While WeakKeyDictionary has no official documentation whehter it preserves order or not, I assume CPython would not drop that de-facto feature without deprecation.

Reply:
Order preservation isn't really sufficient.
The mapping is storing {ax: [sibling0, sibling1, ...], ...} and what matters is that the list of siblings is sorted by the order in which they themselves occur in the mapping.  So we *could* not explicitly store the index and instead sort the siblings by `key=list(self._mapping).index` (taking advantage of order preservation) before returning them but that's less efficient.